beam-sync.md

```diff
@@ -0,0 +1,187 @@
+
+Below are some optimizations that Trinity has found effective. You may notice the theme that they all provide ways to look up more data in parallel. We would much rather be limited by total peer bandwidth across all connected peers, rather than limited by the round-trip latency of a single peer.
+
+#### Optimization 1: Address Batching
+
+Before starting EVM execution, we can identify several key addresses that will be needed during the block: transaction recipients, transaction senders, and the header's coinbase.
+
+Critically, the trie nodes for these addresses can be collected in parallel. Rather than requesting a single trie node at a time, you can walk down one layer of the trie for all the addresses at once. Say you have ~200 addresses to collect, your `GetNodeData` might include up to 200 trie nodes at once on the final layer. This is a dramatic improvement, lessening the sensitivity to peer latency.
+
+#### Optimization 2: Parallel Execution
+
+The flow diagram above shows block imports as serial, so that child blocks aren't executed until the parent is complete. We want to improve on that with Parallel Execution. We start executing child blocks before the parent has completed executing. Some implications:
+
+- The JIT-data limit of only requesting one trie node at a time applies per-block, so they can be requested in parallel across all pending blocks.
+- The block count lag determines how many pending blocks are available. If you are right at the tip, parallelism drops to 1. The benefit of previewing grows as you lag and shrinks as you catch up. It is great to catch up faster, the further you are behind, but it can make Beam Sync feel a bit "stutter-y".
+- Parallel execution "unnecessarily" retrieves state for child blocks, if the parent block writes that state. The data would have been available locally, if the client had waited for the parent to finish. This adds strain to the network. In practice, though, many blocks are largely independent in state access, so the waste is small.
+
+#### Optimization 3: Speculative Execution
+
+Many transactions included in the same block access independent parts of the state. So we can speculatively execute the transactions out-of-order. To speculatively execute, we treat each transaction as if it was the first & only transaction in the block. The transaction may fail, or execute on a different code path, in which case we get no benefit. But for the many state-independent transactions, we can correctly collect the required JIT-data for each transaction in parallel.
+
+Note: if the same block includes multiple transactions from the same sender, then we group them together to run in serial. If you don't group them together, then the 2nd to nth transaction from the sender would always fail the nonce check, and provide no benefit.
+
+There are probably other games you could play to predict transactions that depend on each other, and serialize them. But it hasn't been explored yet.
```

Comment:
You might move this under an separate header since it isn't necessarily part of this one, but a sort of "final thoughts"

Reply:
Ah, I guess I need to connect it better. The idea is that speculative execution might be more likely to successfully follow the correct code path if you serialize the transactions a bit more aggressively. Say, you could serialize all transactions that have the same recipient. (instead of just the ones that have the same sender)